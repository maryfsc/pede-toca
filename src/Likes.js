import React from 'react';
import { Button, Badge } from 'reactstrap';

class Likes extends React.Component {
  
  constructor(props) {
    super(props);
    this.state = {
      likes: 0
    }
  }

  _handleLikes() {    
    this.setState({ likes: this.state.likes + 1 });
  };

  render() {
    return (
      <div className="like-wrapper">
        <Button color="secondary" onClick={() => this._handleLikes()}> 
          Like<Badge color="secondary">{this.state.likes}</Badge>
        </Button>
      </div>
    )
  }
}

export default Likes;

/*
1) What it renders?
  -a button with a badge (span) that contains Likes and a counter for them, starting with 0;

2) Does the component renders different things under different circumstances?
  -No, it's always a button with a badge.

3) When passing a function as a prop, does it calls it? What does it does?
  -Yes, it should make a callback call to update the state;

4) What happens when the user interacts with my component?
  -When the user clicks it, it should update the like count accordingly.

Other contracts:
  - a div is always rendered, and it contains all other elements;
  - there is a button and inside of it, a span with like counts;?
  */
