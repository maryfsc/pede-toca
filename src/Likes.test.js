import React from 'react';
import TestRenderer from 'react-test-renderer';
import Likes from './Likes';

test('Renders a button and badge with like count', () => {
    const createTest = TestRenderer.create(<Likes />);

    let tree = createTest.toJSON();
    
    const testLikes = createTest.root;
    
    tree = createTest.toJSON(); // <-- atualiza o snapshot?
    expect(testLikes.findByType("button")).toBeTruthy();  // espera encontrar um button e retorne true
    expect(testLikes.findByType("span")).toBeTruthy();  // espera encontrar um span ('badge') e retorne true
    
    // tree.props.onClick(); // <-- isso não deveria dar trigger no evento para teste? retorna 'is not a function' (pq é um callback, será?)
    // tree = createTest.toJSON(); // atualiza snapshot após trigger do evento
    // expect(testLikes.children).toBe(1); // ????? sei não, viado

    expect(tree).toMatchSnapshot();    
});

/*
Summary of React Tests:    
  Find your Component Contract first
  Decide which constraints are worth testing and which aren’t
  Prop types are not worth testing
  Inline styles are usually not worth testing
  The components you render and what props you give them are important to test
  Don’t test things that are not the concern of your component
*/

/* Dúvidas, muitas dúvidas!!
  0) Pelo que entendi, mock vai apenas "simular" algum componente ou chamada na hora do teste, é isso?
  1) Quando usar mock? Como decidir isso?
  2) O quanto mock é confiável na hora de fazer os testes? Não é misleading?
  3) Como decidir qual lib de testes usar? Como aprender a usar os métodos da lib escolhida? É confusoooo
  4) Dá pra usar duas ou mais libs pra testar? Isso é uma boa prática?
  5) Como testar funções callback?
  6) Como mockar uma função? Tem um método do jest pra isso pelo que entendi, mas não entendi como funciona
*/