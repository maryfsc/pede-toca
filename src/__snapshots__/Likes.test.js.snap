// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a button and badge with like count 1`] = `
<div>
  <button
    aria-label={null}
    className="btn btn-secondary"
    onClick={[Function]}
    type="button"
  >
    Like
    <span
      className="badge badge-secondary"
    >
      0
    </span>
  </button>
</div>
`;
